packages/node/src/nodeClient.ts

```diff
@@ -46,35 +47,50 @@ export class NodeClient implements Client<Options> {
     // Check if there's 0 events, flush is not needed.
     const arrayLength = this._events.length;
     if (arrayLength === 0) {
-      return { status: Status.Success, statusCode: 200 };
+      return SKIPPED_RESPONSE;
+    }
+
+    // Reset the response listeners and pull them out.
+    const responseListeners = this._responseListeners;
+    this._responseListeners = [];
+
+    try {
+      const eventsToSend = this._events.splice(0, arrayLength);
+      const response = await this._transportWithRetry.sendEventsWithRetry(eventsToSend);
+      responseListeners.forEach(({ resolve }) => resolve(response));
+      return response;
+    } catch (err) {
+      responseListeners.forEach(({ reject }) => reject(err));
+      throw err;
     }
-    const eventsToSend = this._events.splice(0, arrayLength);
-    return this._transportWithRetry.sendEventsWithRetry(eventsToSend);
   }
 
   /**
    * @inheritDoc
    */
-  public logEvent(event: Event): void {
+  public logEvent(event: Event): Promise<Response> {
     if (this._options.optOut === true) {
-      return;
+      return Promise.resolve(SKIPPED_RESPONSE);
     }
 
     this._annotateEvent(event);
-    // Add event to unsent events queue.
-    this._events.push(event);
 
-    if (this._events.length >= this._options.maxCachedEvents) {
-      // # of events exceeds the limit, flush them.
-      this.flush();
-    } else {
-      // Not ready to flush them and not timing yet, then set the timeout
-      if (this._flushTimer === null) {
-        this._flushTimer = setTimeout(() => {
-          this.flush();
-        }, this._options.uploadIntervalInSec * 1000);
+    return new Promise((resolve, reject) => {
+      // Add event to unsent events queue.
+      this._events.push(event);
+      this._responseListeners.push({ resolve, reject });
+      if (this._events.length >= this._options.maxCachedEvents) {
```

Comment:
nit: maybe this should be `>` since maxCachedEvents would imply that you can get up to but not beyond that limit.